File: C/Maths/freivald.c
// Given three matrices A, B and C, find if C is a product of A and B.
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to multiply a matrix by a vector
void matrixVectorMultiply(const int **matrix, const int *vector, int *result, int rows, int cols)
{
    for (int i = 0; i < rows; ++i)
    {
        result[i] = 0;
        for (int j = 0; j < cols; ++j)
        {
            result[i] += matrix[i][j] * vector[j];
        }
    }
}

// Function to generate a random vector with entries 0 or 1
void generateRandomVector(int *vector, int size)
{
    srand(time(NULL));

    for (int i = 0; i < size; ++i)
    {
        vector[i] = rand() % 2;
    }
}

// Function to input a matrix from the user
void inputMatrix(int **matrix, int rows, int cols)
{
    printf("Enter matrix elements row by row:\n");
    for (int i = 0; i < rows; ++i)
    {
        printf("Row %d: ", i + 1);
        for (int j = 0; j < cols; ++j)
        {
            scanf("%d", &matrix[i][j]);
        }
    }
}

int main()
{
    int n, m, k;
    printf("Enter the number of rows for matrix A: ");
    scanf("%d", &n);
    printf("Enter the number of columns for matrix A (the no. of column of A = no. of Rows in matrix B) ");
    scanf("%d", &m);
    int **A = (int **)malloc(n * sizeof(int *));
    for (int i = 0; i < n; ++i)
    {
        A[i] = (int *)malloc(m * sizeof(int));
    }
    inputMatrix(A, n, m);

    printf("Enter the number of columns for matrix B: ");
    scanf("%d", &k);
    int **B = (int **)malloc(m * sizeof(int *));
    for (int i = 0; i < m; ++i)
    {
        B[i] = (int *)malloc(k * sizeof(int));
    }
    inputMatrix(B, m, k);

    int **C = (int **)malloc(n * sizeof(int *));
    for (int i = 0; i < n; ++i)
    {
        C[i] = (int *)malloc(k * sizeof(int));
    }
    printf("For Matrix C \n");
    inputMatrix(C, n, k);

    int *x = (int *)malloc(m * sizeof(int));
    generateRandomVector(x, m);

    int *temp1 = (int *)malloc(n * sizeof(int));
    int *temp2 = (int *)malloc(k * sizeof(int));

    matrixVectorMultiply(B, x, temp1, m, k);
    matrixVectorMultiply(A, temp1, temp2, n, m);

    matrixVectorMultiply(C, x, temp1, n, k);

    int isCorrect = 1;
    for (int i = 0; i < n; ++i)
    {
        if (temp1[i] - temp2[i] != 0)
        {
            isCorrect = 0;
            break;
        }
    }

    if (isCorrect)
    {
        printf("Product is correct.\n");
    }
    else
    {
        printf("Product is incorrect.\n");
    }

    // Free allocated memory
    for (int i = 0; i < n; ++i)
    {
        free(A[i]);
    }
    free(A);

    for (int i = 0; i < m; ++i)
    {
        free(B[i]);
    }
    free(B);

    for (int i = 0; i < n; ++i)
    {
        free(C[i]);
    }
    free(C);

    free(x);
    free(temp1);
    free(temp2);

    return 0;
}

File: Java/Maths/freivald.java

// Given three matrices A, B and C, find if C is a product of A and B.
import java.util.Arrays;
import java.util.Random;
import java.util.Scanner;

class freivald {
    static int n; // Number of rows in matrix A and C
    static int m; // Number of columns in matrix A, number of rows in matrix B
    static int k; // Number of columns in matrix B and C

    // Function to perform matrix-vector multiplication
    static int[] matrixVectorMultiply(int[][] matrix, int[] vector) {
        int[] result = new int[matrix.length];
        Arrays.fill(result, 0);
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                result[i] += matrix[i][j] * vector[j];
            }
        }
        return result;
    }

    // Function to check if the product of A and B equals C
    static String checkProduct(int[][] A, int[][] B, int[][] C) {
        Random random = new Random();
        int[] x = new int[m]; // vector length should be m

        for (int i = 0; i < m; i++) {
            x[i] = random.nextInt(2); // Generating random vector x with 0s and 1s
        }

        int[] y = matrixVectorMultiply(A, matrixVectorMultiply(B, x)); // y = A * (B * x)
        int[] z = matrixVectorMultiply(C, x); // z = C * x

        for (int i = 0; i < n; i++) {
            if (y[i] - z[i] != 0) {
                return "Product is incorrect.";
            }
        }
        return "Product is correct.";
    }

    // Function to read matrix values from user input
    static int[][] readMatrixFromInput(int rows, int cols) {
        Scanner scanner = new Scanner(System.in);
        int[][] matrix = new int[rows][cols];

        System.out.println("Enter matrix values row-wise:");
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = scanner.nextInt();
            }
        }
        return matrix;
    }

    // Driver code
    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of rows in matrix A and C: ");
        n = scanner.nextInt();
        System.out.print("Enter the number of columns in matrix A and rows in matrix B: ");
        m = scanner.nextInt();
        System.out.print("Enter the number of columns in matrix B and C: ");
        k = scanner.nextInt();

        int[][] A = readMatrixFromInput(n, m);
        int[][] B = readMatrixFromInput(m, k);
        int[][] C = readMatrixFromInput(n, k);

        System.out.println(checkProduct(A, B, C));
    }
}
